analytics/transform.go

```diff
@@ -191,22 +192,116 @@ func (event ParsedEvent) GetValue(field string) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
-	if field == "contexts" || field == "derived_contexts" || field == "unstruct_event" {
-		// TODO: DRY HERE TOO?
-		output := make(map[string]interface{})
+
+	return kvPairs, nil
+}
+
+// GetValue returns the value for a provided atomic field, without processing the rest of the event.
+// For unstruct_event, it returns a map of only the data for the unstruct event.
+func (event ParsedEvent) GetValue(field string) (interface{}, error) {
+	kvPairs, err := event.getParsedValue(field)
+	if err != nil {
+		return nil, err
+	}
+	return kvPairs[0].Value, nil
+}
+
+// GetUnstructEventValue returns the value for a provided atomic field inside an event's unstruct_event field
+func (event ParsedEvent) GetUnstructEventValue(path ...interface{}) (interface{}, error) {
+	field := "unstruct_event"
+
+	kvPairs, err := event.getParsedValue(field)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the key/value pairs of the event field into a map
+	eventMap := make(map[string]interface{})
+	for _, pair := range kvPairs {
+		eventMap[pair.Key] = pair.Value
+	}
+
+	j, err := json.Marshal(eventMap)
+	if err != nil {
+		return nil, err
+	}
+
+	b := make([]interface{}, len(path))
+	for idx := range path {
+		b[idx] = path[idx]
+	}
+
+	el := json.Get(j, b...)
+	return el.GetInterface(), nil
+}
+
+// GetContextValue returns the value for a provided atomic field inside an event's contexts or derived_contexts
+func (event ParsedEvent) GetContextValue(contextName string, path ...interface{}) (interface{}, error) {
+	contextNames := []string{`contexts`, `derived_contexts`}
+	var contexts []interface{}
+	for _, c := range contextNames {
+		kvPairs, err := event.getParsedValue(c)
+		if err != nil && err.Error() != EmptyFieldErr {
+			return nil, err
+		}
+		// extract the key/value pairs of the event path into a map
+		eventMap := make(map[string]interface{})
 		for _, pair := range kvPairs {
-			output[pair.Key] = pair.Value
+			eventMap[pair.Key] = pair.Value
 		}
-		return output, nil
+		contexts = append(contexts, eventMap)
 	}
-	return kvPairs[0].Value, nil
 
+	var output []map[string]interface{}
+	b := make([]interface{}, len(path))
+	for idx := range path {
+		b[idx] = path[idx]
+	}
+
+	// iterate through all contextNames and extract the requested path to the output slice
+	for _, ctx := range contexts {
+		for key, contextSlice := range ctx.(map[string]interface{}) {
+			if key == contextName {
+				for _, ctxValues := range contextSlice.([]interface{}) {
+					ctxValuesMap := ctxValues.(map[string]interface{})
+					// output whole context if path is not defined
+					if len(path) == 0 {
+						output = append(output, ctxValuesMap)
+						continue
+					}
+					j, err := json.Marshal(ctxValuesMap)
+					if err != nil {
+						return nil, err
+					}
+					el := json.Get(j, b...)
+					if el.LastError() == nil {
+						var parsedPath string
+						for _, p := range path {
```

Comment:
This part I'm not sure I understand just from reading. Let's cover it on a call.